compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -397,6 +406,170 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
+
+            BLangType bLangTypeOne = constituentTypes.get(0);
+            BType typeOne = symResolver.resolveTypeNode(bLangTypeOne, env);
+            if (typeOne.tag != TypeTags.ERROR && typeOne.tag != TypeTags.READONLY) {
+                dlog.error(bLangTypeOne.pos, INVALID_ERROR_INTERSECTION, bLangTypeOne);
+                continue;
+            }
+
+            BLangType bLangTypeTwo = constituentTypes.get(1);
+            BType typeTwo = symResolver.resolveTypeNode(bLangTypeTwo, env);
+            if (typeTwo.tag != TypeTags.ERROR && typeTwo.tag != TypeTags.READONLY) {
+                dlog.error(bLangTypeTwo.pos, INVALID_ERROR_INTERSECTION, bLangTypeTwo);
+                continue;
+            }
+
+            // Collecting typeIds.
+            if (typeOne.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeOne);
+            }
+            if (typeTwo.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeTwo);
+            }
+
+            BType potentialIntersectionType = types.getTypeIntersection(typeOne, typeTwo);
+
+            boolean isValidIntersection = true;
+            for (int i = 2; i < constituentTypes.size(); i++) {
+                BLangType bLangType = constituentTypes.get(i);
+                BType bType = symResolver.resolveTypeNode(bLangType, env);
+
+                if (bType.getKind() != TypeKind.ERROR && typeTwo.tag != TypeTags.READONLY) {
+                    dlog.error(bLangType.pos, INVALID_ERROR_INTERSECTION, bLangType);
+                    isValidIntersection = false;
+                    break;
+                }
+
+                if (bType.tag != TypeTags.READONLY) {
+                    populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) bType);
+                }
+
+                potentialIntersectionType = types.getTypeIntersection(potentialIntersectionType, bType);
+                if (potentialIntersectionType == symTable.semanticError) {
+                    dlog.error(typeDefinition.typeNode.pos, INVALID_INTERSECTION_TYPE, typeDefinition.typeNode);
+                    isValidIntersection = false;
+                    break;
+                }
+            }
+
+            if (!isValidIntersection) {
+                continue;
+            }
+
+            if (isDistinctType) {
+                ((BErrorType) potentialIntersectionType).typeIdSet = BTypeIdSet.from(
+                        potentialIntersectionType.tsymbol.pkgID, typeDefinition.name.value, isPublicType,
+                        secondaryTypeIds);
+            }
+
+            LinkedHashSet<BType> constituentBTypes = new LinkedHashSet<>() {{
+                add(typeOne);
+                add(typeTwo);
+            }};
+            BIntersectionType intersectionType = definePotentialIntersectionType((BErrorType) potentialIntersectionType,
```

Comment:
Crashes when `potentialIntersectionType` is `symTable.semanticError` due to one error type being invalid.

```ballerina
type Detail record {| // closed, invalid
    string x;
|};

type DetailTwo record {
    string y;
};

type ErrorOne error<Detail>;

type ErrorTwo error<DetailTwo>;

type IntersectionError ErrorOne & ErrorTwo;
```

```log
[2020-11-11 12:52:48,270] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BErrorType (org.wso2.ballerinalang.compiler.semantics.model.types.BType and org.wso2.ballerinalang.compiler.semantics.model.types.BErrorType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BErrorType (org.wso2.ballerinalang.compiler.semantics.model.types.BType and org.wso2.ballerinalang.compiler.semantics.model.types.BErrorType are in unnamed module of loader 'app')
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineErrorIntersection(SymbolEnter.java:479)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineConstructs(SymbolEnter.java:373)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:287)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:249)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.definePackage(SymbolEnter.java:242)
	at org.wso2.ballerinalang.compiler.PackageLoader.define(PackageLoader.java:464)
	at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:375)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:221)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:101)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:109)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:55)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:297)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at org.ballerinalang.tool.Main.main(Main.java:57)
```

Reply:
Fixed